### PR TITLE
Refactor/dialogue controller

### LINKED ISSUE #59: Enable and Disable controls based on system state
## Description

When the pause menu or evidence menu is open, the dialogue in the background should no longer respond to input.

## Dev guide

Disable the object responsible for the input when any menu is open, and enable it again when closing said menu.

## Acceptance Criteria

This story will be finished when the aformentioned functionality is implemented, added to the standard inclusion prefab, and testable in a unity scene.

### LINKED ISSUE #114: [BUG] Cannot present evidence when PRESENT_EVIDENCE action is called
## Describe the bug
When PRESENT_EVIDENCE is called the evidence menu appears but you cannot present evidence.

## Steps To Reproduce
Find a script with the PRESENT_EVIDENCE action.
Try to select evidence to present.

## Expected behavior
You should be able to select evidence to present. Presenting the correct evidence should continue the story. Presenting the wrong evidence should give a failure state.

## Additional context
Presenting evidence only works in cross examination mode. It requires an Ink choice. HOWEVER when PRESENT_EVIDENCE is called it opens the evidence menu which prevents the story from continuing so the choice can never be reached, meaning you cannot select evidence.
In the screenshot the story pauses on PRESENT_EVIDENCE so the choice is not reached.
Screenshot 2021-12-30 at 03 28 47

### LINKED ISSUE #177: [BUG] DialogueController starts playing before BGScenes are loaded
## Describe the bug
BGScenes are created in the Start method of BGScenesList. However DialogueController starts playing in its start method, and relies on BGScenes being loaded, or the &SCENE action will tell BGScenesList to load a BGScene that has not been created yet. 

## Steps To Reproduce

Open Inky-TestScene and run a narrative script where the first action is a &SCENE action e.g., MissingScenesDemo. You may or may not get the following error from BGScenesList.

Screenshot 2022-02-04 at 04 19 11

Order of execution varies within Unity so it may work, it may not.

## Expected behavior
The order of execution must be NarrativeScriptPlaylist's Awake method -> BGSceneList's Start method -> DialogueController's Start method.

## Additional context
If you can do this _without_ setting the order of script execution in Unity that would be good. Personally I feel setting the script execution order is a bit of a dirty fix and is very hard to debug if the order of execution is changed for whatever reason. I feel it'd be better to write the code in a way that indicates the order in which it is run.

### PR DESCRIPTION
**Pull request by PhoebeMitchell**
## Summary

This PR aims to shorten and humblify DialogueController. DialogueConstroller's functionality is now handled by NarrativeScriptPlayer (monobehaviour) and StoryPlayer (testable humble object). In the process #114 should be fixed, and evidence can now be presented using the ink macro `-> Present(evidenceName) ->`. #59 should also be implemented as part of this refactor. Menus now set an active InputModule (previously UnityInputSystemInterface) when they open and close.

This PR also attempts to fix script execution order problems by adding a `Game` monobehaviour which runs scripts in the correct order in its start method. Most of the "controller" objects now have a reference to this monobehaviour.

When fixing the tests I found ActionDecoder seemed to have reverted back to how it was before ActionDecoderBase was implemented. I assumed this was a problem with the recent repo structure changes, so it now inherits from ActionDecoderBase again.

## Testing instructions

Run the tests. Does the game work? Try all the scenes.

## Additional information

- `[ ]` Changes UI
- `[ ]` Introduces new feature
- `[ ]` Removes existing feature
- `[x]` Has associated resource: 
  
closes #114 
closes #177 
closes #59

### CODE REVIEW

**Review by notexplosive — DISMISSED** [review]

File: unity-ggjj/Assets/Scripts/Menu/ChoiceMenu/ChoiceMenu.cs [hunk 1]
```diff
@@ -52,10 +58,16 @@ public void InitialiseChoiceMenu(List<Choice> choiceList)
                 _menu.SelectInitialButton();
             }
             menuItem.Text = choice.text;
-            ((Button)menuItem.Selectable).onClick.AddListener(() => _onChoiceClicked.Invoke(choice.index));
+            ((Button)menuItem.Selectable).onClick.AddListener(() => OnChoiceClicked(choice.index));
         }
     }
 
+    private void OnChoiceClicked(int choiceIndex)
```
Comment (lyndsiWilliams) [inline unity-ggjj/Assets/Scripts/Menu/ChoiceMenu/ChoiceMenu.cs:62]: Nit: This method is missing a summary ✨

File: unity-ggjj/Assets/Scripts/NarrativeScript/NarrativeScriptPlayer/NarrativeScriptPlayer.cs [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    {
+        GetComponent<Game>().NarrativeScriptPlayer = this;
+    }
+
+    public void Continue()
+    {
+        StoryPlayer.Continue();
+    }
+
+    public void HandleChoice(int choiceIndex)
```
Comment (lyndsiWilliams) [inline unity-ggjj/Assets/Scripts/NarrativeScript/NarrativeScriptPlayer/NarrativeScriptPlayer.cs:33]: Nit: All of the methods in this file are missing summaries ✨

File: unity-ggjj/Assets/Scripts/NarrativeScript/StoryPlayer.cs [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+                _subStory.GameMode = value;
+            }
+            else
+            {
+                _gameMode = value;
+            }
+        }
+    }
+
+    public StoryPlayer(INarrativeScriptPlaylist narrativeScriptPlaylist, IAppearingDialogueController appearingDialogueController, IActionDecoder actionDecoder, IChoiceMenu choiceMenu)
```
Comment (lyndsiWilliams) [inline unity-ggjj/Assets/Scripts/NarrativeScript/StoryPlayer.cs:70]: Nit: All of the methods in this file are missing summaries ✨

File: unity-ggjj/Assets/Scripts/UI/AppearingDialogueController/AppearingDialogueController.cs [hunk 4]
```diff
@@ -109,6 +104,18 @@ public void PrintText(string text)
         _printCoroutine = StartCoroutine(PrintTextCoroutine(startingIndex));
     }
 
+    public void StopPrintingText()
```
Comment (lyndsiWilliams) [inline unity-ggjj/Assets/Scripts/UI/AppearingDialogueController/AppearingDialogueController.cs:104]: Nit: This method is missing a summary ✨

**Review by lyndsiWilliams — COMMENTED** [review]
A few summary nits (not merge blocking), also I've got 3 tests consistently failing locally (possibly merge blocking, unless I'm doing something wrong again 😅): `CanNavigateToMultiplePages`, `MouseShouldHighlightEvidenceMenuItems`, `NavigationButtonsCanBeClicked`

[Github gist of errors](https://gist.github.com/lyndsiWilliams/bf9bb8b8389b2e7feba7140894b5f5b5)

**Review by notexplosive — APPROVED** [review]

**Pull request closed without merging**